src/main/java/org/javarosa/core/model/FormDef.java

```diff
@@ -1441,6 +1455,41 @@ public void initialize(boolean newInstance, boolean isCompletedInstance,
         }
     }
 
+    private void tryAttachingRemoteInstances(RemoteInstanceFetcher remoteInstanceFetcher) throws RemoteInstanceFetcher.RemoteInstanceException {
+        ArrayList<DataInstance> replacedInstances = new ArrayList<>();
+        Enumeration<DataInstance> instances = getNonMainInstances();
+        while (instances.hasMoreElements()) {
+            DataInstance instance = instances.nextElement();
+            if (instance instanceof ExternalDataInstance &&
+                    instance.getRoot() == null &&
+                    ((ExternalDataInstance)instance).getRemoteUrl() != null) {
+                try {
+                    ExternalDataInstance externalDataInstance = (ExternalDataInstance)instance;
+                    ExternalDataInstance newExternalDataInstance = remoteInstanceFetcher.getRemoteDataInstance(instance.getInstanceId(),
+                            externalDataInstance.useCaseTemplate(),
+                            new URI(externalDataInstance.getRemoteUrl()));
+                    if (newExternalDataInstance != null && newExternalDataInstance.getRoot() != null) {
+                        replacedInstances.add(newExternalDataInstance);
+                    }
+                } catch (UnfullfilledRequirementsException | XmlPullParserException |
+                        InvalidStructureException | IOException | URISyntaxException e) {
+                    String errorMessage;
+                    if (e instanceof URISyntaxException) {
```

Comment:
why not split these out into their own catch blocks instead of doing type checking?

Reply:
I was trying to avoid the repetition on line below  `throw new RemoteInstanceFetcher.RemoteInstanceException(errorMessage, e.getCause());` but open to either styles if this doesn't look neat.